### PR TITLE
feat: arrow join methods

### PR DESCRIPTION
**Pull request by FBruzzesi**
## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [x] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue #372 

## Checklist

- [x] Code follows style guide (ruff)
- [x] Tests added 
- [x] Documented the changes

## If you have comments or can explain your changes, please do so below.

Implements various join types for arrow and `fill_null` (as it is used in one test)

I am mostly trusting our tests at this point :)

### CODE REVIEW

File: tests/frame/join_test.py [hunk 1]
```diff
@@ -35,15 +35,11 @@ def test_inner_join_single_key(constructor: Any) -> None:
...
-    if "pyarrow_table" in str(constructor):
-        request.applymarker(pytest.mark.xfail)
-
+def test_cross_join(constructor: Any) -> None:
     data = {"a": [1, 3, 2]}
     df = nw.from_native(constructor(data))
-    result = df.join(df, how="cross")  # type: ignore[arg-type]
-
-    expected = {"a": [1, 1, 1, 3, 3, 3, 2, 2, 2], "a_right": [1, 3, 2, 1, 3, 2, 1, 3, 2]}
+    result = df.join(df, how="cross").sort("a", "a_right")  # type: ignore[arg-type]
```
Comment (FBruzzesi) [inline tests/frame/join_test.py:41]: Without sorting, the result is flaky as order is not guaranteed

**Comment by MarcoGorelli** [comment]
amazing!

https://github.com/vega/altair/pull/3487 should fix the CI failure

**Comment by binste** [comment]
> amazing!
> 
> [vega/altair#3487](https://github.com/vega/altair/pull/3487) should fix the CI failure

FYI, that PR just got merged

**Review by MarcoGorelli — APPROVED** [review]
not long til we move "pyarrow" into the "full api support" category, exciting 🎉

**Pull request merged**